src/jit/morph.cpp

```diff
@@ -7836,6 +7836,14 @@ void Compiler::fgMorphTailCall(GenTreeCall* call, void* pfnCopyArgs)
     JITDUMP("fgMorphTailCall (before):\n");
     DISPTREE(call);
 
+    // The runtime requires that we perforrm a null check on the `this` argument before tail
+    // calling to a virtual dispatch stub. This requirement is a consequence of limitations in the runtime's
+    // ability to map an AV to a NullReferenceException if the AV occurs in a dispatch stub.
```

Comment:
`... that has unmanaged caller`. The runtime is able to map an AV to a NullReferenceException exception in a general case. It is just the unmanaged caller case that is problematic.